File: LEETCODE/LEETCODE-CPP/547.cpp
//547. Number of Provinces
//https://leetcode.com/problems/number-of-provinces/description/

//Method-1 using DFS
class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n=isConnected.size(), count=0;
        vector<bool>visited(n,false);

        for(int i=0;i<n;i++){
            if(visited[i]==false){
                visited[i]=true;
                count++;
                helper(isConnected,visited,i,n);
            }
        }

        return count;
    }

    void helper(vector<vector<int>>& isConnected, vector<bool>& visited, int curr, int n){
        for(int i=0;i<n;i++){
            if(visited[i]==false && isConnected[curr][i]==1){
                visited[i]=true;
                helper(isConnected,visited,i,n);
            }
        }
    }
};

File: LEETCODE/LEETCODE-PYTHON/547.py
# 547. Number of Provinces
# https://leetcode.com/problems/number-of-provinces/description/

# Method-1 using DFS
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n= len(isConnected)
        visited = [False]*n
        count=0

        def helper(x):
            visited[x]=True

            for j in range(n):
                if isConnected[x][j]==1 and visited[j]==False:
                    helper(j)

        for i in range(n):
            if visited[i]== False:
                count+=1
                helper(i)

        return count